### PR TITLE
Set defaults of init containers.

### LINKED ISSUE #38542: imagePullPolicy is required on Init Containers within a Pod template
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

* initContainers imagePullPolicy
* initContainers required value

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.3", GitCommit:"35fd1363821b8432ccde55632cf9ead4c79eddbb", GitTreeState:"clean", BuildDate:"2016-12-08T21:58:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-beta.3", GitCommit:"35fd1363821b8432ccde55632cf9ead4c79eddbb", GitTreeState:"clean", BuildDate:"2016-12-08T21:54:19Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**: cluster/kube-up.sh
- **Others**:

**What happened**:

When trying to create either a ReplicaSet or Replication Controller whose Pod template contains an Init Container, the validation fails saying `imagePullPolicy` is a required value.

```
The ReplicaSet "nginx" is invalid: spec.template.spec.initContainers[0].imagePullPolicy: Required value
```

```
The ReplicationController "nginx" is invalid: spec.template.spec.initContainers[0].imagePullPolicy: Required value
```

**What you expected to happen**:

The create should succeed and the Init Container should receive a default `imagePullPolicy`.

**How to reproduce it** (as minimally and precisely as possible):

Try to create either of the following:

**ReplicaSet**

```yaml
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  name: nginx
spec:
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        pod.beta.kubernetes.io/init-containers: '[
            {
                "name": "install",
                "image": "busybox",
                "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
                "volumeMounts": [
                    {
                        "name": "workdir",
                        "mountPath": "/work-dir"
                    }
                ]
            }
        ]'
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
        volumeMounts:
        - name: workdir
          mountPath: /usr/share/nginx/html
      dnsPolicy: Default
      volumes:
      - name: workdir
        emptyDir: {}
```

**Replication Controller**

```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: nginx
spec:
  replicas: 3
  selector:
    app: nginx
  template:
    metadata:
      labels:
        app: nginx
      annotations:
        pod.beta.kubernetes.io/init-containers: '[
            {
                "name": "install",
                "image": "busybox",
                "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
                "volumeMounts": [
                    {
                        "name": "workdir",
                        "mountPath": "/work-dir"
                    }
                ]
            }
        ]'
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
        volumeMounts:
        - name: workdir
          mountPath: /usr/share/nginx/html
      dnsPolicy: Default
      volumes:
      - name: workdir
        emptyDir: {}
```

**Anything else do we need to know**:

If I add `"imagePullPolicy": "IfNotPresent",` to the Init Container in the ReplicaSet or Replication Controller Pod template, the create succeeds.

When I create an Init Container without `imagePullPolicy` in a StatefulSet Pod template, it succeeds.

When I create an Init Container without `imagePullPolicy` in a bare Pod, it succeeds:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  annotations:
    pod.beta.kubernetes.io/init-containers: '[
        {
            "name": "install",
            "image": "busybox",
            "command": ["wget", "-O", "/work-dir/index.html", "http://kubernetes.io/index.html"],
            "volumeMounts": [
                {
                    "name": "workdir",
                    "mountPath": "/work-dir"
                }
            ]
        }
    ]'
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
    volumeMounts:
    - name: workdir
      mountPath: /usr/share/nginx/html
  dnsPolicy: Default
  volumes:
  - name: workdir
    emptyDir: {}
```

### PR DESCRIPTION
**Pull request by k82cn**
fixes #38542

### CODE REVIEW

File: pkg/api/v1/helpers.go [hunk 1]
```diff
@@ -454,3 +455,70 @@ type NodeResources struct {
 	// Capacity represents the available resources of a node
 	Capacity ResourceList `protobuf:"bytes,1,rep,name=capacity,casttype=ResourceList,castkey=ResourceName"`
 }
+
+func (rc *ReplicationController) DecodeNestedObjects(d runtime.Decoder) error {
```
Comment (smarterclayton) [inline pkg/api/v1/helpers.go]: Hrm, this wasn't what this method was intended to do.  It's not technically wrong, but I'll need to think about it.

In any case, the methods don't belong in this file, they'll need to be in something specific like decode.go.

**Comment by hongchaodeng** [comment]
Can you help explain why it didn't set before? And How are you going to fix?

**Comment by k82cn** [comment]
@hongchaodeng , in https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/serializer/versioning/versioning.go#L96 , it'll decode, set default and the convert; but the old code decodes init container when coverting, which missed "set default" (the old code only call `SetDefault_Container` which is not enough).

I'm going to implement `runtime.NestedObjectDecoder` for the kind who will include init container; so the init container objects are here when setting default & converting.

But as @smarterclayton mentioned in comments, it maybe not a good place for init container.

File: pkg/api/v1/helpers.go (hunk 1 above)
Comment (k82cn) [inline pkg/api/v1/helpers.go]: @smarterclayton , any suggestion on what's original intention of `runtime.NestedObjectDecoder`? Did not find example implementation in master branch.

I'll move it (e.g. decode.go) when we got finial decision :).

**Comment by smarterclayton** [comment]
Add a test that shows the break before.

**Comment by k82cn** [comment]
@smarterclayton, the test for pull image is still here, I just removed the code in `defaults.go` ( https://github.com/k82cn/kubernetes/blob/init_container_defaults/pkg/api/v1/defaults_test.go#L234 ).

Do you think that's the right direction to handle this issue? If so, I'll continue to handle it.

**Comment by k82cn** [comment]
@smarterclayton , thanks for your comments at https://groups.google.com/forum/#!topic/kubernetes-dev/wy2JLRzugPI :

>Maybe we should talk about that on the issue, but since we're promoting init containers soon and we are probably moving away from annotations as alpha fields as well it probably wouldn't be used that much.

I'd like to set defaults for init containers manually in this tickets; because this is not issue after we moved annotation to alpha fields.
I copy the set default codes for init containers from `zz_generated.defaults.go` to `defaults.go`. I'm going to more test cases: imagePullPolicy, apiVersion and so on.

**Comment by k82cn** [comment]
@smarterclayton , added test cases for init container's defults. PTAL :).

**Comment by k82cn** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/api/v1/helpers.go, line 459 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39713#-KaB4YDoMmI5QH1GW1LO-r1-459:-KaH3Olhvfa0bmjxMxQY:biksd76) ([raw file](https://github.com/kubernetes/kubernetes/blob/639d87971d746be6c698d090a3177651650a3dad/pkg/api/v1/helpers.go#L459)):*
<details><summary><i>Previously, k82cn (Klaus Ma) wrote…</i></summary><blockquote>

@smarterclayton , any suggestion on what's original intention of `runtime.NestedObjectDecoder`? Did not find example implementation in master branch.

I'll move it (e.g. decode.go) when we got finial decision :).
</blockquote></details>

Updated PR that did not use `runtime.NestedObjectDecoder`.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39713)*

File: pkg/api/v1/defaults.go [hunk 2]
```diff
@@ -349,3 +349,49 @@ func SetDefaults_RBDVolumeSource(obj *RBDVolumeSource) {
 		obj.Keyring = "/etc/ceph/keyring"
 	}
 }
+
+// TODO(k82cn): remove this defaulter when init container graduated to beta.
+func SetDefaults_InitContainers(initContainers []Container) {
```
Comment (k82cn) [inline pkg/api/v1/defaults.go]: Sure, let me try it out :).

**Comment by k82cn** [comment]
Just update current status: I'm working on generating decode nested code; just learned gengo's code recently :).

**Comment by k82cn** [comment]
The WIP patches at: https://github.com/k82cn/gengo/tree/k8s_39713/examples/nestedobj-gen , will create PR when it's ready.

**Comment by smarterclayton** [comment]
Sorry, I didn't mean "generate a decoder".  I meant use the existing defaulter-gen code to generate a top level defaulter for containers that was a single method you could call.

**Comment by k82cn** [comment]
@smarterclayton , ok :). generating initContainer default seems more reasonable.

**Comment by k82cn** [comment]
Updated the WIP PR:

- [ ] check nil for pointers
- [ ] fix failed test in `serialize_test.go`

**Comment by k82cn** [comment]
@smarterclayton / @thockin thanks for your comments at https://github.com/kubernetes/gengo/pull/38 , I'd like to move the discussion here to get the finial decision.

Here's what we had tried:
1. Add `SetDefaults_InitContainers` manually as current PR
1. Add `gengo` to generate `DecodeNestedObjects` as https://github.com/kubernetes/gengo/pull/38
1. Try generate `SetDefaults_InitContainers` by defaulter-gen; but default-gen only consider top object, e.g. Pod

I prefer to option 1 (current PR) for this issue: after moved init-continer to field, there's no such issue; the other two options is hard to maintain and should roll-back after moved init-container to field.

**Comment by thockin** [comment]
I think I agree

**Comment by k82cn** [comment]
@smarterclayton , any comments?

**Comment by smarterclayton** [comment]
We're about to merge promotion of init container to GA.  At that point we
can rebase this and verify the beta behavior fix.

On Feb 1, 2017, at 7:28 PM, Klaus Ma <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> , any comments?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/39713#issuecomment-276828339>,
or mute the thread

**Comment by k82cn** [comment]
@smarterclayton , so we're move annotation to field? any PR about it?

**Comment by derekwaynecarr** [comment]
@k82cn see https://github.com/kubernetes/kubernetes/pull/38382

**Comment by k82cn** [comment]
@derekwaynecarr / @smarterclayton , thanks for info; #38382 does not fix this issue because the initContainers was override by annotation.

And here's the root cause of this issue for you reference. After #38382 , the initContainers in annotation will override field when converting.

> In https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/serializer/versioning/versioning.go#L96 , it'll decode, set default and the convert; but the old code decodes init container when coverting, which missed "set default" (the old code only call SetDefault_Container which is not enough).

**Comment by k82cn** [comment]
refer to #41004

**Comment by k82cn** [comment]
ping @derekwaynecarr / @smarterclayton , any suggestion to the previous comments?

**Comment by k82cn** [comment]
@derekwaynecarr / @smarterclayton , do you think we should include this PR in 1.6 release?

File: pkg/api/v1/conversion.go [hunk 3]
```diff
@@ -625,10 +622,7 @@ func Convert_v1_Pod_To_api_Pod(in *Pod, out *api.Pod, s conversion.Scope) error
 		// back to the caller.
 		in.Spec.InitContainers = values
 		// Call defaulters explicitly until annotations are removed
-		for i := range in.Spec.InitContainers {
-			c := &in.Spec.InitContainers[i]
-			SetDefaults_Container(c)
-		}
+		SetDefaults_InitContainers(in.Spec.InitContainers)
```
Comment (liggitt) [inline pkg/api/v1/conversion.go]: is there a reason we can't just do this?
```
tmpPod := &Pod{Spec:PodSpec{InitContainers:values}}
SetObjectDefaults_Pod(tmpPod)
in.Spec.InitContainers = tmpPod.Spec.InitContainers
```

the copied SetDefaults_InitContainers is going to drift

**Comment by liggitt** [comment]
it looks like SetDefaults_Pod is missing iterating over InitContainers as well... shouldn't it do both?

**Comment by k82cn** [comment]
@liggitt , it's handled in `SetObjectDefaults_Pod` (`zz_generated.defaults.go`).

File: pkg/api/v1/conversion.go (hunk 3 above)
Comment (k82cn) [inline pkg/api/v1/conversion.go]: wow, that's great :). Let's me try it.

**Comment by liggitt** [comment]
so is this comment inaccurate?
```
func SetDefaults_Pod(obj *Pod) {
	// If limits are specified, but requests are not, default requests to limits
	// This is done here rather than a more specific defaulting pass on ResourceRequirements
	// because we only want this defaulting semantic to take place on a Pod and not a PodTemplate
	for i := range obj.Spec.Containers {
		...
	}
}
```

**Comment by smarterclayton** [comment]
At this point, now that init containers are GA, I'd just suggest users use
init containers in GA form.  I'm sorry to say that after all the hard work
you put in this, but fixing something for beta use cases is riskier and
more involved than just switching to GA init containers.

**Comment by k82cn** [comment]
> At this point, now that init containers are GA, I'd just suggest users use
init containers in GA form. 

Are we going to put up with this issue for beta features? If so, I'd suggest to only commit test part for init containers.

**Comment by liggitt** [comment]
regardless of whether we end up applying defaults to the version pulled from the annotation, I think SetDefaults_Pod needs updating for init containers. Opened https://github.com/kubernetes/kubernetes/pull/42310 to address that separately

**Comment by k82cn** [comment]
Updated PR according to liggitt 's comments to avoid new set default func.

**Comment by k82cn** [comment]
@k8s-bot cvm gce e2e test this

File: pkg/api/v1/conversion.go [hunk 4]
```diff
@@ -480,10 +480,9 @@ func Convert_v1_PodTemplateSpec_To_api_PodTemplateSpec(in *PodTemplateSpec, out
 		in.Spec.InitContainers = values
 
 		// Call defaulters explicitly until annotations are removed
-		for i := range in.Spec.InitContainers {
-			c := &in.Spec.InitContainers[i]
-			SetDefaults_Container(c)
-		}
+		tmpPod := &Pod{Spec: PodSpec{InitContainers: values}}
+		SetObjectDefaults_Pod(tmpPod)
```
Comment (liggitt) [inline pkg/api/v1/conversion.go]: hmm, this won't catch all cases... things like this propagate other parts of the pod spec down into the containers:
```
	if obj.HostNetwork {
		defaultHostNetworkPorts(&obj.Containers)
		defaultHostNetworkPorts(&obj.InitContainers)
	}
```

Comment (liggitt) [inline pkg/api/v1/conversion.go]: that's actually the only special case I can find. I'd be ok with setting the tmpPod.Spec.HostNetwork as well, and calling it a day. @smarterclayton, what's the timeline on dropping support for initcontainer annotations altogether?

Comment (liggitt) [inline pkg/api/v1/conversion.go]: I suppose 1.6 kubelets are the first to report initcontainer status in fields, so in 1.8 we can drop the annotations and use only the fields

Comment (k82cn) [inline pkg/api/v1/conversion.go]: yes, I think we need to continue to make it work until 1.8, although annotation is deprecated.

Comment (k82cn) [inline pkg/api/v1/conversion.go]: Done :).

**Comment by liggitt** [comment]
do we also need to default InitContainerStatuses in Convert_v1_PodStatusResult_To_api_PodStatusResult?

**Comment by k82cn** [comment]
Nop, no set default func for `InitContainerStatuses`. And I think this should be updated by kubelet :).

File: pkg/api/v1/defaults_test.go [hunk 5]
```diff
@@ -232,55 +232,418 @@ func TestSetDefaultReplicationControllerReplicas(t *testing.T) {
...
+			name: "FieldRef",
+			rc: v1.ReplicationController{
+				Spec: v1.ReplicationControllerSpec{
+					Template: &v1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{
+							Annotations: map[string]string{
+								"pod.beta.kubernetes.io/init-containers": `
+                                [
+                                    {
+                                    "Name": "fun",
```
Comment (liggitt) [inline pkg/api/v1/defaults_test.go]: these are incorrect... there should never be upper-case keys in API json objects... did you serialize internal API objects to obtain these?

File: staging/src/k8s.io/client-go/pkg/api/v1/conversion.go [hunk 6]
```diff
@@ -481,10 +481,12 @@ func Convert_v1_PodTemplateSpec_To_api_PodTemplateSpec(in *PodTemplateSpec, out
...
 		// Call defaulters explicitly until annotations are removed
-		for i := range in.Spec.InitContainers {
-			c := &in.Spec.InitContainers[i]
-			SetDefaults_Container(c)
-		}
+		tmpPod := &Pod{Spec: PodSpec{
+			HostNetwork:    in.Spec.HostNetwork,
+			InitContainers: values,
+		}}
+		SetObjectDefaults_Pod(tmpPod)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/pkg/api/v1/conversion.go]: should be `SetObjectDefaults_PodTemplate`

File: pkg/api/v1/defaults_test.go (hunk 5 above)
Comment (k82cn) [inline pkg/api/v1/defaults_test.go]: Good catch ! but it seems work in this test :(, I also add some instrument code to make sure the case is run.

In `roundTrip`, it uses `runtime.Decode` & `api.Scheme.Convert` to convert the objs.

Comment (k82cn) [inline pkg/api/v1/defaults_test.go]: let me check why, and update codes accordingly :).

Comment (k82cn) [inline pkg/api/v1/defaults_test.go]: It seems go support such kind of deserialization, here's the example to show the result:

source code: https://gist.github.com/k82cn/5081f64d160dc166e29dc2d0ac4e057b
output:

```
macpro:golearn klaus$ ./golearn
[{Name:Platypus Order:Monotremata} {Name:Quoll Order:Dasyuromorphia}]
{"name":"Platypus","order":"Monotremata"}
```

go version: `go version go1.7.5 darwin/amd64`

File: staging/src/k8s.io/client-go/pkg/api/v1/conversion.go (hunk 6 above)
Comment (k82cn) [inline staging/src/k8s.io/client-go/pkg/api/v1/conversion.go]: Good catch ! Done :).

File: pkg/api/v1/defaults_test.go (hunk 5 above)
Comment (liggitt) [inline pkg/api/v1/defaults_test.go]: go json does, ugorji does not. don't use upper case in examples or tests

Comment (k82cn) [inline pkg/api/v1/defaults_test.go]: got that. I will update it:)

Comment (k82cn) [inline pkg/api/v1/defaults_test.go]: Done

File: pkg/api/v1/conversion.go [hunk 7]
```diff
@@ -604,10 +610,16 @@ func Convert_v1_Pod_To_api_Pod(in *Pod, out *api.Pod, s conversion.Scope) error
 		// back to the caller.
 		in.Spec.InitContainers = values
 		// Call defaulters explicitly until annotations are removed
-		for i := range in.Spec.InitContainers {
-			c := &in.Spec.InitContainers[i]
-			SetDefaults_Container(c)
+		tmpPodTemp := &PodTemplate{
```
Comment (liggitt) [inline pkg/api/v1/conversion.go]: Convert_v1_Pod_To_api_Pod should still use a &Pod{}

Comment (liggitt) [inline pkg/api/v1/conversion.go]: Convert_v1_Pod_To_api_Pod should still use a &Pod{}

Comment (k82cn) [inline pkg/api/v1/conversion.go]: ack ; sorry for the mis-understand.

**Comment by liggitt** [comment]
this seems reasonable to me

**Comment by k82cn** [comment]
@liggitt , what's our next step? Merge to 1.5 or master?

**Comment by liggitt** [comment]
> @liggitt , what's our next step? Merge to 1.5 or master?

approval by @smarterclayton, then master, then any picks we want

**Comment by k82cn** [comment]
ping  @smarterclayton for approve :).

**Comment by k82cn** [comment]
ping @smarterclayton for approve :).

The annotation will sunset until 1.8; there's still some people using 1.5 who keep reporting this issue. So I'd suggest to merge this PR to master & 1.5.

**Comment by k82cn** [comment]
@k8s-bot kops aws e2e test this

**Comment by k82cn** [comment]
@liggitt / @smarterclayton , the cherrypick-candidate label was removed; should we cherrypick for 1.5 & 1.6?

**Comment by phanidileep** [comment]
Was this included in 1.5?

**Pull request merged**